### PR TITLE
editoast: adapt infra to new model (retrieve and delete)

### LINKED ISSUE #3717: editoast : refacto Infra with the new model

### PR DESCRIPTION
**Pull request by younesschrifi**
Close #3717

### CODE REVIEW

File: editoast/src/main.rs [hunk 1]
```diff
@@ -185,53 +193,76 @@ async fn generate(
...
 
-fn import_railjson(
+async fn import_railjson(
     args: ImportRailjsonArgs,
     pg_config: PostgresConfig,
 ) -> Result<(), Box<dyn Error + Send + Sync>> {
     let railjson_file = File::open(args.railjson_path)?;
+    let manager = ConnectionManager::<PgConnection>::new(pg_config.url());
+    let pool = Data::new(Pool::builder().max_size(1).build(manager).unwrap());
     let conn = &mut PgConnection::establish(&pg_config.url()).expect("Error while connecting DB");
```
Comment (Castavo) [inline editoast/src/main.rs]: Isn't it weird to use `conn` and `pool` at the same time ?

Comment (younesschrifi) [reply editoast/src/main.rs]: Tbh, it triggered me too, maybe we should ask @flomonster abt that

Comment (flomonster) [reply editoast/src/main.rs]: Indeed we should only use the pool here.

File: editoast/src/main.rs [hunk 2]
```diff
@@ -296,24 +351,24 @@ mod tests {
...
-            Ok(())
-        });
-    }
+    // pub fn test_transaction(fn_test: fn(&mut PgConnection)) {
+    //     let mut conn = PgConnection::establish(&PostgresConfig::default().url()).unwrap();
+    //     conn.test_transaction::<_, Error, _>(|conn| {
+    //         fn_test(conn);
+    //         Ok(())
+    //     });
+    // }
```
Comment (flomonster) [inline editoast/src/main.rs]: Dead code

File: editoast/src/models/infra.rs [hunk 3]
```diff
@@ -94,75 +97,78 @@ impl Infra {
...
             .set(dsl::version.eq(new_version.to_string()))
             .get_result::<Infra>(conn)
         {
             Ok(infra) => Ok(infra),
-            Err(DieselError::NotFound) => Err(InfraApiError::NotFound { infra_id: self.id }.into()),
+            Err(DieselError::NotFound) => Err(InfraApiError::NotFound {
+                infra_id: self.id.unwrap(),
+            }
+            .into()),
             Err(err) => Err(err.into()),
```
Comment (flomonster) [inline editoast/src/models/infra.rs]: We could use the `Self::update` function.

File: editoast/src/models/infra.rs [hunk 4]
```diff
@@ -94,75 +97,78 @@ impl Infra {
...
             .set(dsl::modified.eq(Utc::now().naive_utc()))
             .get_result::<Infra>(conn)
         {
             Ok(infra) => Ok(infra),
-            Err(DieselError::NotFound) => Err(InfraApiError::NotFound { infra_id: self.id }.into()),
+            Err(DieselError::NotFound) => Err(InfraApiError::NotFound {
+                infra_id: self.id.unwrap(),
+            }
+            .into()),
             Err(err) => Err(err.into()),
```
Comment (flomonster) [inline editoast/src/models/infra.rs]: We can use the `Self::update` function with something like that.

```rust
let infra = Self::default();
infra.update(conn, self.id.unwrap());
```

File: editoast/src/models/infra.rs [hunk 5]
```diff
@@ -94,75 +97,78 @@ impl Infra {
...
             + 1;
 
-        match update(dsl::osrd_infra_infra.filter(dsl::id.eq(self.id)))
+        match update(dsl::osrd_infra_infra.filter(dsl::id.eq(self.id.unwrap())))
             .set(dsl::version.eq(new_version.to_string()))
             .get_result::<Infra>(conn)
         {
             Ok(infra) => Ok(infra),
-            Err(DieselError::NotFound) => Err(InfraApiError::NotFound { infra_id: self.id }.into()),
+            Err(DieselError::NotFound) => Err(InfraApiError::NotFound {
```
Comment (flomonster) [inline editoast/src/models/infra.rs]: To match other models we should use Errors in views only.

File: editoast/src/models/infra.rs [hunk 6]
```diff
@@ -94,75 +97,78 @@ impl Infra {
...
-            Err(err) => Err(err.into()),
-        }
+    pub async fn persist(self, railjson: RailJson, db_pool: Data<DbPool>) -> Result<Infra> {
+        block::<_, Result<_>>(move || {
+            if railjson.version != RAILJSON_VERSION {
+                return Err(RailjsonError::WrongVersion(railjson.version).into());
+            }
+            let mut conn = db_pool.get()?;
+            let infra = self.create_conn(&mut conn)?;
+            let infra_id = infra.id.unwrap();
```
Comment (flomonster) [inline editoast/src/models/infra.rs]: This should be in the transaction.

File: editoast/src/views/infra/mod.rs [hunk 7]
```diff
@@ -60,6 +64,32 @@ pub fn routes() -> impl HttpServiceFactory {
...
+#[serde(deny_unknown_fields)]
+pub struct InfraForm {
+    pub name: String,
+}
+
+impl From<InfraForm> for Infra {
+    fn from(infra: InfraForm) -> Self {
+        Infra {
+            name: Some(infra.name),
+            owner: Some(uuid!("00000000-0000-0000-0000-000000000000")),
```
Comment (flomonster) [inline editoast/src/views/infra/mod.rs]: Maybe we should set this value as default.

File: editoast/Cargo.toml [hunk 8]
```diff
@@ -15,11 +15,13 @@ clap = { version = "4.1", features = ["derive", "env"] }
 colored = "2.0.0"
 chrono = { version = "0.4", features = ["serde"] }
 derivative = "2.2.0"
+uuid = { version = "1.3.0", features = ["v4"] }
```
Comment (flomonster) [inline editoast/Cargo.toml]: To use the same version as the one used by diesel, we should only lock the major version. Checkout https://github.com/diesel-rs/diesel/blob/9fea8dfb98389acde9bd5c1666131da1c8c8ce24/diesel/Cargo.toml#L26

```suggestion
uuid = { version = "1", features = ["v4"] }
```

**Review by flomonster — CHANGES_REQUESTED** [review]
Thanks for this PR. The infra utility functions can use more from the models abstraction.

File: editoast/src/models/infra.rs [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+    /// Lock or unlock the infra whether `lock` is true or false
+    pub fn set_locked(&self, lock: bool, conn: &mut PgConnection) -> Result<Self> {
+        match update(dsl::osrd_infra_infra.filter(dsl::id.eq(self.id.unwrap())))
+            .set(dsl::locked.eq(lock))
+            .get_result::<Infra>(conn)
+        {
+            Ok(infra) => Ok(infra),
+            Err(err) => Err(err.into()),
+        }
+    }
```
Comment (flomonster) [inline editoast/src/models/infra.rs]: This is not needed. We can call the `update` function.

File: editoast/src/models/infra.rs [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+        infra.version = new_version;
+        let infra = infra.update_conn(conn, infra_id)?.unwrap();
+        Ok(infra)
+    }
+
+    pub fn bump_generated_version(&self, conn: &mut PgConnection) -> Result<Self> {
+        let infra = self.clone();
+        let infra = infra.update_conn(conn, self.id.unwrap())?.unwrap();
+        Ok(infra)
+    }
```
Comment (flomonster) [inline editoast/src/models/infra.rs]: Same here we can use the update function.

File: editoast/src/models/infra.rs [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+impl Infra {
+    pub fn retrieve_for_update(conn: &mut PgConnection, infra_id: i64) -> Result<Infra> {
+        let infra = dsl::osrd_infra_infra
+            .for_update()
+            .find(infra_id)
+            .first(conn)?;
+        Ok(infra)
+    }
+
+    pub fn list(conn: &mut PgConnection) -> Vec<Infra> {
```
Comment (flomonster) [inline editoast/src/models/infra.rs]: We can use the `List` trait here.

File: editoast/src/views/infra/mod.rs (hunk 7 above)
Comment (flomonster) [inline editoast/src/views/infra/mod.rs]: We can use https://docs.rs/uuid/latest/uuid/struct.Uuid.html#method.nil

```suggestion
            owner: Some(Uuid::nil()),
```

File: editoast/src/views/infra/mod.rs [hunk 12]
```diff
@@ -147,24 +181,20 @@ async fn list(db_pool: Data<DbPool>) -> Result<Json<Vec<Infra>>> {
...
-    })
-    .await
-    .unwrap()
+    let infra_id = infra.into_inner();
+
+    let infra = match Infra::retrieve(db_pool.clone(), infra_id).await? {
+        Some(infra) => infra,
+        None => return Err(InfraApiError::NotFound { infra_id }.into()),
+    };
+    Ok(Json(infra))
```
Comment (flomonster) [inline editoast/src/views/infra/mod.rs]:
```suggestion
    match Infra::retrieve(db_pool.clone(), infra_id).await? {
        Some(infra) => Ok(Json(infra)),
        None => Err(InfraApiError::NotFound { infra_id }.into()),
    }
```

File: editoast/src/views/infra/mod.rs [hunk 13]
```diff
@@ -442,35 +487,32 @@ pub mod tests {
...
-        assert_eq!(response.status(), StatusCode::NOT_FOUND);
     }
 
     #[actix_test]
     async fn infra_rename() {
         let app = create_test_service().await;
         let infra: Infra =
             call_and_read_body_json(&app, create_infra_request("infra_rename_test")).await;
-
+        println!("ok");
```
Comment (flomonster) [inline editoast/src/views/infra/mod.rs]:
```suggestion
```

**Review by flomonster — CHANGES_REQUESTED** [review]
Thanks for this PR! I have reviewed it without testing it yet

File: editoast/src/models/projects.rs [hunk 14]
```diff
@@ -297,9 +297,19 @@ pub mod test {
 
         let projects = ProjectWithStudies::list(pool.clone(), 1, 25, Ordering::NameDesc)
             .await
-            .unwrap();
-        let name = projects.results.first().unwrap().project.name.clone();
-        assert_eq!(name, Some("test".into()));
+            .unwrap()
+            .results;
+
+        for i in 1..projects.len() {
```
Comment (flomonster) [inline editoast/src/models/projects.rs]:
```suggestion
        for (p1, p2) in projects.iter().zip(projects.iter().skip(1)) {
```

File: editoast/src/models/scenario.rs [hunk 15]
```diff
@@ -331,20 +340,31 @@ pub mod test {
...
         let scenarios =
             ScenarioWithCountTrains::list(pool.clone(), 1, 25, (study_id, Ordering::NameDesc))
                 .await
-                .unwrap();
-        let name = scenarios.results.first().unwrap().scenario.name.clone();
-        assert_eq!(name, Some("test".into()));
+                .unwrap()
+                .results;
+
+        for i in 1..scenarios.len() {
```
Comment (flomonster) [inline editoast/src/models/scenario.rs]: Same here

**Review by flomonster — CHANGES_REQUESTED** [review]

**Review by clarani — CHANGES_REQUESTED** [review]
Bug on the cartographie page

**Review by clarani — COMMENTED** [review]
Actually, the bug is already fixed ! We can just change the openApi to specify the required properties :)

**Review by anisometropie — DISMISSED** [review]

File: editoast/Cargo.toml [hunk 16]
```diff
@@ -15,12 +15,14 @@ clap = { version = "4.1", features = ["derive", "env"] }
...
 chrono = { version = "0.4", features = ["serde"] }
 derivative = "2.2.0"
+uuid = { version = "1", features = ["v4"] }
 diesel = { version = "2.0", features = [
     "r2d2",
     "postgres",
     "serde_json",
     "chrono",
     "uuid",
+
```
Comment (flomonster) [inline editoast/Cargo.toml]:
```suggestion
```

File: editoast/src/models/study.rs [hunk 17]
```diff
@@ -235,15 +235,20 @@ pub mod test {
...
-                .unwrap();
-        let name = studies.results.first().unwrap().study.name.clone();
-        assert_eq!(name, Some("test".into()));
+                .unwrap()
+                .results;
+        for (p1, p2) in studies.iter().zip(studies.iter().skip(1)) {
+            let name_1 = p1.study.name.as_ref().unwrap().to_lowercase();
+            let name_2 = p2.study.name.as_ref().unwrap().to_lowercase();
+            println!("{name_1}");
+            println!("{name_2}");
```
Comment (flomonster) [inline editoast/src/models/study.rs]:
```suggestion
```

**Review by flomonster — DISMISSED** [review]
LGTM. Two minor changes. I didn't test the whole application.

**Review by clarani — APPROVED** [review]
LGTM for the part front, tested in local (carto and editor pages)

**Review by flomonster — APPROVED** [review]

**Pull request merged**